sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -494,7 +506,27 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case DateType =>
       buildCast[Int](_, d => null)
     case TimestampType =>
-      buildCast[Long](_, t => timestampToLong(t).toShort)
+      buildCast[Long](_, t => {
+        val longValue = timestampToLong(t)
+        if (!failOnIntegerOverflow || longValue == longValue.toShort) {
+          longValue.toShort
+        } else {
+          throw new ArithmeticException(s"Casting $t to Short causes overflow.")
+        }
+      })
+    case x: NumericType if failOnIntegerOverflow =>
+      b =>
+        val intValue = try {
+          x.exactNumeric.asInstanceOf[Numeric[Any]].toInt(b)
```

Comment:
Why do you cast it into int once?

Reply:
The trait `Numeric` doesn't have the method `toInt`. Before this code change, the value is also casted to int.

Reply:
Ah, I see.

Reply:
We cannot check the valid value range in a single place instead of the current two checks in line 520 and 525?

Reply:
Well, we can do it by match it case by case. Then the code is a bit long. Casting to short/byte should be minor usage. Also, The previous code also cast to `Int` before cast to `Short`.